functions/common/aws-s3-common/src/main/java/org/springframework/cloud/fn/common/aws/s3/CompatibleStorageAmazonS3Configuration.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright 2020-2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.fn.common.aws.s3;
+
+import com.amazonaws.auth.AWSCredentialsProvider;
+import com.amazonaws.client.builder.AwsClientBuilder.EndpointConfiguration;
+import com.amazonaws.services.s3.AmazonS3;
+import com.amazonaws.services.s3.AmazonS3ClientBuilder;
+
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.boot.context.properties.EnableConfigurationProperties;
+import org.springframework.cloud.aws.core.region.RegionProvider;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.context.annotation.Primary;
+
+/**
+ * @author Timo Salm
+ */
+@Configuration
+@EnableConfigurationProperties(AmazonS3Properties.class)
```

Comment:
Mark this class with `@AutoconfigureBefore(AmazonS3Configuration.class)` and you don't need that `@Primary` any more!

Reply:
Results in a failing testAmazonS3ConfigurationForS3CompatibleStorage Test ->  `org.springframework.beans.factory.NoUniqueBeanDefinitionException: No qualifying bean of type 'com.amazonaws.services.s3.AmazonS3' available: expected single matching bean but found 2: amazonS3,compatibleStorageAmazonS3`

Reply:
(Please, study how to do multi-line code snippet in GH comments: https://docs.github.com/en/github/writing-on-github/creating-and-highlighting-code-blocks)

If you see that behavior then something is wrong. 
We definitely should have only a single `AmazonS3` bean.
Probably that `AmazonS3Configuration` comes without a `@ConditionalOnMissingBean`. Time to fix that?

Reply:
So it looks like without the @Primary the AmazonS3 Bean created by the spring-cloud-starter-aws dependency will always be the default (and also the AmazonS3 Bean in AmazonS3Configuration will never be used).

Reply:
As far as I know the `spring-cloud-starter-aws` doesn't create `AmazonS3`. Please, elaborate more.
Thanks

Reply:
Turns out the `ContextResourceLoaderAutoConfiguration` creates for us S3 client bean and that one cannot be overridden. Probably time to talk to Spring Cloud AWS team...